apps/ejabberd/src/ejabberd_c2s.erl

```diff
@@ -1369,6 +1186,157 @@ handle_info(Info, StateName, StateData) ->
     ?ERROR_MSG("Unexpected info: ~p", [Info]),
     fsm_next_state(StateName, StateData).
 
+handle_routed(<<"presence">>, From, To, Packet, StateData) ->
+    handle_routed_presence(From, To, Packet, StateData);
+handle_routed(<<"broadcast">>, _From, _To, Packet, StateData) ->
+    handle_routed_broadcast(Packet, StateData);
+handle_routed(<<"iq">>, From, To, Packet, StateData) ->
+    handle_routed_iq(From, To, Packet, StateData);
+handle_routed(<<"message">>, From, To, Packet, StateData) ->
+    case privacy_check_packet(StateData, From, To, Packet, in) of
+	allow ->
+	    {true, Packet#xmlel.attrs, StateData};
+	deny ->
+	    {false, Packet#xmlel.attrs, StateData}
+    end;
+handle_routed(_, _From, _To, Packet, StateData) ->
+    {true, Packet#xmlel.attrs, StateData}.
+
+handle_routed_iq(From, To, Packet = #xmlel{attrs = Attrs}, StateData) ->
+    case jlib:iq_query_info(Packet) of
+	%% TODO: Support for mod_last / XEP-0012. Can we move it to the respective module?
```

Comment:
I don't thinks. The hook in its current version doesn't contain state date which is used by the mod_last functionality (as far as I remember).
This could be possible via hook only if we add new type of hook extracting the c2s state if needed and passing it as an argument to the handler.